lib/streamlit/elements/widgets/data_editor.py

```diff
@@ -817,15 +817,31 @@ def data_editor(
 
         # Convert the user provided column config into the frontend compatible format:
         column_config_mapping = process_config_mapping(column_config)
-        apply_data_specific_configs(
-            column_config_mapping, data_df, data_format, check_arrow_compatibility=True
-        )
+
+        # Deactivate editing for columns that are not compatible with arrow
+        for column_name, column_data in data_df.items():
+            if dataframe_util.is_colum_type_arrow_incompatible(column_data):
+                update_column_config(
+                    column_config_mapping, column_name, {"disabled": True}
+                )
+                # Convert incompatible type to string
+                data_df[column_name] = column_data.astype("string")
+
+        apply_data_specific_configs(column_config_mapping, data_format)
 
         # Fix the column headers to work correctly for data editing:
         _fix_column_headers(data_df)
-        # Temporary workaround: We hide range indices if num_rows is dynamic.
-        # since the current way of handling this index during editing is a bit confusing.
-        if isinstance(data_df.index, pd.RangeIndex) and num_rows == "dynamic":
+
+        if not isinstance(data_df.index, pd.RangeIndex):
+            # If the index is not a range index, we will configure it as required
+            # since the user is required to provide a (unique) value for editing.
+            update_column_config(
+                column_config_mapping, INDEX_IDENTIFIER, {"required": True}
+            )
+        elif num_rows == "dynamic":
```

Comment:
it looks like you could merge the `elif` branch with the `if hide_index` [check below](https://github.com/streamlit/streamlit/pull/9205/files#diff-e591171c22c1c88a8244dbf9fd17e4f30b2d9d9cd93f8bac5479cd1ecd992448R849) since `hidden` will be overridden by that check anyways (which seems to make the comment also a little bit inaccurate).
```python
if hide_index is not None and num_rows == "dynamic":
```

Reply:
I think it is not as easy to merge since: 1) this should only happen if the index is a range index and 2) the user should have the possibility to overwrite; and 3) if none of these conditions are True, it shouldn't set hidden at all. 

I pushed a small change to improve readability a bit.